homeassistant/components/screenlogic/__init__.py

```diff
@@ -0,0 +1,202 @@
+    SL_GATEWAY_PORT,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_IP_ADDRESS, CONF_PORT, CONF_SCAN_INTERVAL
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .config_flow import async_discover_gateways_by_unique_id, name_for_mac
+from .const import DEFAULT_SCAN_INTERVAL, DISCOVERED_GATEWAYS, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = ["switch", "sensor", "binary_sensor", "water_heater"]
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Screenlogic component."""
+    domain_data = hass.data[DOMAIN] = {}
+    domain_data[DISCOVERED_GATEWAYS] = await async_discover_gateways_by_unique_id(hass)
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Screenlogic from a config entry."""
+    mac = entry.unique_id
+    # Attempt to re-discover named gateway to follow IP changes
+    discovered_gateways = hass.data[DOMAIN][DISCOVERED_GATEWAYS]
+    if mac in discovered_gateways:
+        connect_info = discovered_gateways[mac]
+    else:
+        _LOGGER.warning("Gateway rediscovery failed.")
+        # Static connection defined or fallback from discovery
+        connect_info = {
+            SL_GATEWAY_NAME: name_for_mac(mac),
+            SL_GATEWAY_IP: entry.data[CONF_IP_ADDRESS],
+            SL_GATEWAY_PORT: entry.data[CONF_PORT],
+        }
+
+    try:
+        gateway = ScreenLogicGateway(**connect_info)
+    except ScreenLogicError as ex:
+        _LOGGER.error("Error while connecting to the gateway %s: %s", connect_info, ex)
+        raise ConfigEntryNotReady from ex
+    except AttributeError as ex:
+        _LOGGER.exception(
+            "Unexpected error while connecting to the gateway %s", connect_info
+        )
+        raise ConfigEntryNotReady from ex
+
+    coordinator = ScreenlogicDataUpdateCoordinator(
+        hass, config_entry=entry, gateway=gateway
+    )
+
+    entities = defaultdict(list)
+
+    await coordinator.async_refresh()
+
+    for circuit in coordinator.data["circuits"]:
+        entities["switch"].append(circuit)
+
+    for sensor in coordinator.data["sensors"]:
+        if sensor == "chem_alarm":
+            entities["binary_sensor"].append(sensor)
+        else:
+            if coordinator.data["sensors"][sensor]["value"] != 0:
+                entities["sensor"].append(sensor)
+
+    for pump in coordinator.data["pumps"]:
+        if (
+            coordinator.data["pumps"][pump]["data"] != 0
+            and "currentWatts" in coordinator.data["pumps"][pump]
+        ):
+            entities["pump"].append(pump)
+
+    for body in coordinator.data["bodies"]:
+        entities["water_heater"].append(body)
```

Comment:
This this data is already accessible and stored in `coordinator.data`, it would be a bit cleaner to have each platform use it directly instead of creating the intermediate data

Reply:
The raw data stored in the coordinator has more fields than are really useful in home assistant. This intermediate data is a filtered down set of just what should be exposed in home assistant. It could be done for each platform in async_setup_entry(), but I felt that is was more readable to have a singular central place to do/see that.

Reply:
Maybe call this `device_data` instead of `entities` as there aren't home assistant entities which is going to trip up future reviews since we have a specific prohibition on storing entities in a shared container.

Reply:
Can definitely rename for clarity.